src/hyperlight_host/src/hypervisor/hyperlight_vm.rs

```diff
@@ -0,0 +1,768 @@
+            outb_handle_fn,
+            mem_access_fn,
+            #[cfg(gdb)]
+            dbg_mem_access_fn,
+        )?;
+
+        Ok(())
+    }
+
+    #[instrument(err(Debug), skip_all, parent = Span::current(), level = "Trace")]
+    fn handle_io(
+        &mut self,
+        port: u16,
+        data: Vec<u8>,
+        outb_handle_fn: OutBHandlerWrapper,
+    ) -> Result<()> {
+        // KVM does not need RIP or instruction length, as it automatically sets the RIP
+
+        let mut padded = [0u8; 4];
+        let copy_len = data.len().min(4);
+        padded[..copy_len].copy_from_slice(&data[..copy_len]);
+        let val = u32::from_le_bytes(padded);
+
+        if data.is_empty() {
+            log_then_return!("no data was given in IO interrupt");
+        } else {
+            outb_handle_fn
+                .try_lock()
+                .map_err(|e| new_error!("Error locking at {}:{}: {}", file!(), line!(), e))?
+                .call(port, val)?;
+        }
+
+        Ok(())
+    }
+
+    fn run(
+        &mut self,
+        hv_handler: Option<HypervisorHandler>,
+        outb_handle_fn: Arc<Mutex<dyn OutBHandlerCaller>>,
+        mem_access_fn: Arc<Mutex<dyn MemAccessHandlerCaller>>,
+        #[cfg(gdb)] dbg_mem_access_fn: DbgMemAccessHandlerWrapper,
+    ) -> Result<()> {
+        loop {
+            match self.vm.run_vcpu() {
+                #[cfg(gdb)]
+                Ok(HyperlightExit::Debug(debug_exit)) => {
+                    match debug_exit {
+                        DebugExit::Debug { dr6, exception } => {
+                            // Handle debug event (breakpoints)
+                            let stop_reason = arch::vcpu_stop_reason(
+                                self.vm.as_mut(),
+                                self.entrypoint,
+                                dr6,
+                                exception,
+                            )?;
+                            if let Err(e) =
+                                self.handle_debug(dbg_mem_access_fn.clone(), stop_reason)
+                            {
+                                log_then_return!(e);
+                            }
+                        }
+                        DebugExit::Interrupt => {
+                            if let Err(e) = self
+                                .handle_debug(dbg_mem_access_fn.clone(), VcpuStopReason::Interrupt)
+                            {
+                                log_then_return!(e);
+                            }
+                        }
+                    }
+                }
+
+                Ok(HyperlightExit::Halt()) => {
+                    break;
+                }
+                Ok(HyperlightExit::IoOut(port, data)) => {
+                    self.handle_io(port, data, outb_handle_fn.clone())?
+                }
+                Ok(HyperlightExit::MmioRead(addr)) => {
+                    #[cfg(crashdump)]
+                    crashdump::crashdump_to_tempfile(self)?;
+
```

Comment:
Currently, on Windows, we have some very useful `debug!`s for different `HyperlightExit`s that capture VCPU state. I think these were removed from `whp.rs`. Would be good to re-add them here, so we have consistent `debug!`s for all drivers.

Reply:
Isn't the crashdump feature enough for this? I can add them back if you prefer though

Reply:
I tend to agree with Dan, debug output might be used when crashdump feature is off, however we should probably not do this in release builds since registers might contain sensitive data